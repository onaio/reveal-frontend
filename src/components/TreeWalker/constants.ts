export const ACTIVE = 'Active';
export const FEATURE = 'Feature';
export const FIND_BY_ID = 'location/findByJurisdictionIds';
export const FIND_BY_PROPERTIES = 'location/findByProperties';
export const LOCATION = 'location';
export const COULDNT_LOAD_CHILDREN = 'Could not load children';
export const COULDNT_LOAD_PARENTS = 'Could not load parents';
export const NOT_AVAILABLE = 'N/A';
