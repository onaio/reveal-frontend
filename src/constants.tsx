// strings
export const MAP = 'map';
export type MAP = typeof MAP;
export const LOCATION = 'Location';
export type LOCATION = typeof LOCATION;
export const PROVINCE = 'Province';
export type PROVINCE = typeof PROVINCE;
export const HOME = 'Home';
export type HOME = typeof HOME;
export const FOCUS_INVESTIGATION = 'Focus Investigation';
export type FOCUS_INVESTIGATION = typeof FOCUS_INVESTIGATION;
export const ACTIVE_FOCUS_INVESTIGATION = 'Active Focus Investigations';
export type ACTIVE_FOCUS_INVESTIGATION = typeof ACTIVE_FOCUS_INVESTIGATION;
export const FOCUS_AREA_INFO = 'Focus Area Information';
export type FOCUS_AREA_INFO = typeof FOCUS_AREA_INFO;
export const FOCUS_INVESTIGATIONS = 'Focus Investigations';
export type FOCUS_INVESTIGATIONS = typeof FOCUS_INVESTIGATIONS;
export const MAP_ID = 'map-1';
export type MAP_ID = typeof MAP_ID;
export const GEOJSON = 'geojson';
export type GEOJSON = typeof GEOJSON;
export const STRINGIFIED_GEOJSON = 'stringified-geojson';
export type STRINGIFIED_GEOJSON = typeof STRINGIFIED_GEOJSON;
export const FOCUS_AREA_HEADER = 'Focus Area';
export type FOCUS_AREA_HEADER = typeof FOCUS_AREA_HEADER;
export const REASON_HEADER = 'Reason';
export type REASON_HEADER = typeof REASON_HEADER;
export const STATUS_HEADER = 'Status';
export type STATUS_HEADER = typeof STATUS_HEADER;
export const CASE_NOTIF_DATE_HEADER = 'Case Notif. Date';
export type CASE_NOTIF_DATE_HEADER = typeof CASE_NOTIF_DATE_HEADER;
export const CASE_CLASSIFICATION_HEADER = 'Case Class.';
export type CASE_CLASSIFICATION_HEADER = typeof CASE_CLASSIFICATION_HEADER;
export const DEFINITIONS = 'Definitions';
export type DEFINITIONS = typeof DEFINITIONS;
export const IN = 'in';
export type IN = typeof IN;
export const DISTRICT = 'District';
export type DISTRICT = typeof DISTRICT;
export const CANTON = 'Canton';
export type CANTON = typeof CANTON;
export const FI_REASON = 'FI Reason';
export type FI_REASON = typeof FI_REASON;
export const FI_STATUS = 'FI Status';
export type FI_STATUS = typeof FI_STATUS;
export const ACTIVE_INVESTIGATION = 'Active Investigation';
export type ACTIVE_INVESTIGATION = typeof ACTIVE_INVESTIGATION;
export const COMPLETE = 'Complete';
export type COMPLETE = typeof COMPLETE;
export const NO = 'No';
export type NO = typeof NO;
export const RESPONSE = 'Response';
export type RESPONSE = typeof RESPONSE;
export const MEASURE = 'Measure';
export type MEASURE = typeof MEASURE;
export const OF = 'of';
export type OF = typeof OF;
export const MARK_AS_COMPLETE = 'Mark as complete';
export type MARK_AS_COMPLETE = typeof MARK_AS_COMPLETE;
export const TARGET = 'Target';
export type TARGET = typeof TARGET;
export const POLYGON = 'Polygon';
export type POLYGON = typeof POLYGON;
export const MULTI_POLYGON = 'MultiPolygon';
export type MULTI_POLYGON = typeof MULTI_POLYGON;
export const POINT = 'Point';
export type POINT = typeof POINT;
export const FEATURE = 'Feature';
export type FEATURE = typeof FEATURE;
export const FEATURE_COLLECTION = 'FeatureCollection';
export type FEATURE_COLLECTION = typeof FEATURE_COLLECTION;
export const MAIN_PLAN = 'main-plan-layer';
export type MAIN_PLAN = typeof MAIN_PLAN;
export const APP = 'APP';
export type APP = typeof APP;
export const NO_GEOMETRIES_RESPONSE = 'Goals have no Geometries';
export type NO_GEOMETRIES_RESPONSE = typeof NO_GEOMETRIES_RESPONSE;

// internal urls
export const LOGIN_URL = '/login';
export type LOGIN_URL = typeof LOGIN_URL;
export const LOGOUT_URL = '/logout';
export type LOGOUT_URL = typeof LOGOUT_URL;
export const HOME_URL = '/';
export type HOME_URL = typeof HOME_URL;
export const IRS_URL = '/irs';
export type IRS_URL = typeof IRS_URL;
export const FI_URL = '/focus-investigation';
export type FI_URL = typeof FI_URL;
export const FI_SINGLE_URL = '/focus-investigation/view';
export type FI_SINGLE_URL = typeof FI_SINGLE_URL;
export const FI_SINGLE_MAP_URL = '/focus-investigation/map';
export type FI_SINGLE_MAP_URL = typeof FI_SINGLE_MAP_URL;
export const FI_HISTORICAL_URL = '/focus-investigation/historical';
export type FI_HISTORICAL_URL = typeof FI_HISTORICAL_URL;

// colors
export const GREEN = 'Green';
export type GREEN = typeof GREEN;
export const YELLOW = 'Yellow';
export type YELLOW = typeof YELLOW;
export const ORANGE = 'Orange';
export type ORANGE = typeof ORANGE;
export const RED = 'Red';
export type RED = typeof RED;

// task action codes
export const RACD_REGISTER_FAMILY_CODE = 'RACD Register Family';
export type RACD_REGISTER_FAMILY_CODE = typeof RACD_REGISTER_FAMILY_CODE;
export const CASE_CONFIRMATION_CODE = 'Case Confirmation';
export type CASE_CONFIRMATION_CODE = typeof CASE_CONFIRMATION_CODE;
export const MOSQUITO_COLLECTION_CODE = 'Mosquito Collection';
export type MOSQUITO_COLLECTION_CODE = typeof MOSQUITO_COLLECTION_CODE;
export const BEDNET_DISTRIBUTION_CODE = 'Bednet Distribution';
export type BEDNET_DISTRIBUTION_CODE = typeof BEDNET_DISTRIBUTION_CODE;
export const LARVAL_DIPPING_CODE = 'Larval Dipping';
export type LARVAL_DIPPING_CODE = typeof LARVAL_DIPPING_CODE;
export const BLOOD_SCREENING_CODE = 'Blood Screening';
export type BLOOD_SCREENING_CODE = typeof BLOOD_SCREENING_CODE;
export const IRS_CODE = 'IRS';
export type IRS = typeof IRS_CODE;
