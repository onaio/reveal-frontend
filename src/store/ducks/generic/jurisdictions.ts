import { get, keyBy, values } from 'lodash';
import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';

/** the reducer name */
export const reducerName = 'GenericJurisdictions';

/** GenericJurisdiction interface */
export interface GenericJurisdiction {
  id: string;
  jurisdiction_depth: number;
  jurisdiction_id: string;
  jurisdiction_name: string;
  jurisdiction_name_path: string[];
  jurisdiction_parent_id: string;
  jurisdiction_path: string[];
  plan_id: string;
}

export interface NamibiaJurisdiction extends GenericJurisdiction {
  lockedfirst: number;
  lockedmopup: number;
  refusalsfirst: number;
  refusalsmopup: number;
  sprayeffectiveness: number;
  structuresfound: number;
  structuressprayed: number;
  targetcoverage: number;
  target_2019: number;
}

export interface ZambiaJurisdiction extends GenericJurisdiction {
  totstruct: number;
  rooms_eligible: number;
  rooms_sprayed: number;
  sprayed_rooms_eligible: number;
  sprayed_rooms_sprayed: number;
  foundstruct: number;
  sprayedstruct: number;
  totareas: number;
  targareas: number;
  targstruct: number;
  perctvisareaseffect: number;
  spraycovtarg: number;
  roomcov: number;
}

export interface ZambiaIRSFocusArea extends GenericJurisdiction {
  totstruct: number;
  rooms_eligible: number;
  rooms_sprayed: number;
  sprayed_rooms_eligible: number;
  sprayed_rooms_sprayed: number;
  foundstruct: number;
  sprayedstruct: number;
  spraycov: number;
  spraytarg: number;
  spraysuccess: number;
  roomcov: number;
}

// actions

/** GENERIC_JURISDICTION_FETCHED action type */
export const GENERIC_JURISDICTIONS_FETCHED =
  'reveal/reducer/IRS/GenericJurisdiction/GENERIC_JURISDICTIONS_FETCHED';

/** GENERIC_JURISDICTION_FETCHED action type */
export const REMOVE_GENERIC_JURISDICTIONS =
  'reveal/reducer/IRS/GenericJurisdiction/REMOVE_GENERIC_JURISDICTIONS';

/** GENERIC_JURISDICTION_FETCHED action type */
export const ADD_GENERIC_JURISDICTION =
  'reveal/reducer/IRS/GenericJurisdiction/ADD_GENERIC_JURISDICTION';

/** interface for fetch GenericJurisdictions action */
interface FetchGenericJurisdictionsAction extends AnyAction {
  objects: { [key: string]: GenericJurisdiction };
  reducerKey: string;
  type: typeof GENERIC_JURISDICTIONS_FETCHED;
}

/** interface for removing GenericJurisdictions action */
interface RemoveGenericJurisdictionsAction extends AnyAction {
  objects: { [key: string]: GenericJurisdiction };
  reducerKey: string;
  type: typeof REMOVE_GENERIC_JURISDICTIONS;
}

/** interface for adding a single GenericJurisdictions action */
interface AddGenericJurisdictionAction extends AnyAction {
  obj: GenericJurisdiction;
  reducerKey: string;
  type: typeof ADD_GENERIC_JURISDICTION;
}

/** Create type for GenericJurisdiction reducer actions */
export type GenericJurisdictionActionTypes =
  | FetchGenericJurisdictionsAction
  | RemoveGenericJurisdictionsAction
  | AddGenericJurisdictionAction
  | AnyAction;

// action creators

/**
 * Fetch Plan Definitions action creator
 * @param {string} reducerKey - they reducer key
 * @param {GenericJurisdiction[]} objList - list of plan definition objects
 */
export const fetchGenericJurisdictions = (
  reducerKey: string = 'GenericJurisdictionsById',
  objList: GenericJurisdiction[] = []
): FetchGenericJurisdictionsAction => ({
  objects: keyBy(
    objList.map(
      (obj: GenericJurisdiction): GenericJurisdiction => {
        /** ensure geojson is parsed */
        if ((obj as any).geojson && typeof (obj as any).geojson === 'string') {
          (obj as any).geojson = JSON.parse((obj as any).geojson);
        }
        /** ensure geometry is parsed */
        if (
          (obj as any).geojson &&
          (obj as any).geojson.geometry &&
          typeof (obj as any).geojson.geometry === 'string'
        ) {
          (obj as any).geojson.geometry = JSON.parse((obj as any).geojson.geometry);
        }
        /** ensure jurisdiction_name_path is parsed */
        if (typeof obj.jurisdiction_name_path === 'string') {
          obj.jurisdiction_name_path = JSON.parse(obj.jurisdiction_name_path);
        }
        /** ensure jurisdiction_path is parsed */
        if (typeof obj.jurisdiction_path === 'string') {
          obj.jurisdiction_path = JSON.parse(obj.jurisdiction_path);
        }
        return obj as GenericJurisdiction;
      }
    ),
    'id'
  ),
  reducerKey,
  type: GENERIC_JURISDICTIONS_FETCHED,
});

/** Reset plan definitions state action creator
 * @param {string} reducerKey - they reducer key
 */
export const removeGenericJurisdictions = (reducerKey: string = 'GenericJurisdictionsById') => ({
  objects: {},
  reducerKey,
  type: REMOVE_GENERIC_JURISDICTIONS,
});

/**
 * Add one Plan Definition action creator
 * @param {GenericJurisdiction} obj - the plan definition object
 * @param {string} reducerKey - they reducer key
 */
export const addGenericJurisdiction = (
  reducerKey: string = 'GenericJurisdictionsById',
  obj: GenericJurisdiction
): AddGenericJurisdictionAction => ({
  obj,
  reducerKey,
  type: ADD_GENERIC_JURISDICTION,
});

// the reducer

/** interface for GenericJurisdiction state */
interface GenericJurisdictionState {
  [key: string]: {
    [key: string]: NamibiaJurisdiction | ZambiaJurisdiction | ZambiaIRSFocusArea;
  };
}

/** immutable GenericJurisdiction state */
export type ImmutableGenericJurisdictionState = GenericJurisdictionState &
  SeamlessImmutable.ImmutableObject<GenericJurisdictionState>;

/** initial GenericJurisdiction state */
const initialState: ImmutableGenericJurisdictionState = SeamlessImmutable({});

/** the GenericJurisdiction reducer function */
export default function reducer(
  state: any = initialState,
  action: GenericJurisdictionActionTypes
): ImmutableGenericJurisdictionState {
  switch (action.type) {
    case GENERIC_JURISDICTIONS_FETCHED:
      if (action.objects) {
        return SeamlessImmutable({
          ...state,
          [action.reducerKey]: { ...state[action.reducerKey], ...action.objects },
        });
      }
      return state;
    case ADD_GENERIC_JURISDICTION:
      if (action.obj as GenericJurisdiction) {
        return SeamlessImmutable({
          ...state,
          [action.reducerKey]: {
            ...state[action.reducerKey],
            [action.obj.id as string]: action.obj,
          },
        });
      }
      return state;
    case REMOVE_GENERIC_JURISDICTIONS:
      return SeamlessImmutable({
        ...state,
        [action.reducerKey]: action.objects,
      });
    default:
      return state;
  }
}

// selectors

/** get GenericJurisdictions by id
 * @param {Partial<Store>} state - the redux store
 * @param {string} reducerKey - they reducer key
 * @param {string} planId - the plan id
 * @param {string} parentId - the parent id
 * @returns {{ [key: string]: GenericJurisdiction }} GenericJurisdictions by id
 */
export function getGenericJurisdictionsById(
  state: Partial<Store>,
  reducerKey: string = 'GenericJurisdictionsById',
  planId: string | null = null,
  parentId: string | null = null
): { [key: string]: GenericJurisdiction } {
  return keyBy(getGenericJurisdictionsArray(state, reducerKey, planId, parentId), 'id');
}

/** get one GenericJurisdiction using its id
 * @param {Partial<Store>} state - the redux store
 * @param {string} reducerKey - they reducer key
 * @param {string} id - the GenericJurisdiction id
 * @returns {GenericJurisdiction|null} a GenericJurisdiction object or null
 */
export function getGenericJurisdictionById(
  state: Partial<Store>,
  reducerKey: string = 'GenericJurisdictionsById',
  id: string
): GenericJurisdiction | null {
  return get((state as any)[reducerName][reducerKey], id) || null;
}

/** get one GenericJurisdiction using its jurisdiction_id
 * @param {Partial<Store>} state - the redux store
 * @param {string} reducerKey - they reducer key
 * @param {string} jurisdictionId - the GenericJurisdiction id
 * @returns {GenericJurisdiction|null} a GenericJurisdiction object or null
 */
export function getGenericJurisdictionByJurisdictionId(
  state: Partial<Store>,
  reducerKey: string = 'GenericJurisdictionsById',
  jurisdictionId: string
): GenericJurisdiction | null {
  return values((state as any)[reducerName][reducerKey]).filter(
    (e: GenericJurisdiction) => e.jurisdiction_id === jurisdictionId
  )[0];
}

/** get an array of GenericJurisdiction objects
 * @param {Partial<Store>} state - the redux store
 * @param {string} reducerKey - they reducer key
 * @param {string} planId - the plan id
 * @param {string} parentId - the parent id
 * @returns {GenericJurisdiction[]} an array of GenericJurisdiction objects
 */
export function getGenericJurisdictionsArray(
  state: Partial<Store>,
  reducerKey: string = 'GenericJurisdictionsById',
  planId: string | null = null,
  parentId: string | null = null
): GenericJurisdiction[] {
  let result = values((state as any)[reducerName][reducerKey]);
  if (planId) {
    result = result.filter((e: GenericJurisdiction) => e.plan_id === planId);
  }
  if (parentId) {
    return result.filter((e: GenericJurisdiction) => e.jurisdiction_parent_id === parentId);
  }
  return result;
}
