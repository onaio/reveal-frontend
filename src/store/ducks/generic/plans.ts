import { get, keyBy, values } from 'lodash';
import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';
import { InterventionType, PlanStatus } from '../plans';

/** the reducer name */
export const reducerName = 'IRSPlans';

/** IRSPlan interface */
export interface IRSPlan {
  plan_date: string;
  plan_effective_period_end: string;
  plan_effective_period_start: string;
  plan_fi_reason: string;
  plan_fi_status: string;
  plan_id: string;
  plan_intervention_type: InterventionType;
  plan_name: string;
  plan_status: PlanStatus;
  plan_title: string;
  plan_version: string;
  jurisdiction_root_parent_ids: string[];
}

// actions

/** IRS_PLAN_FETCHED action type */
export const IRS_PLANS_FETCHED = 'reveal/reducer/IRS/IRSPlan/IRS_PLANS_FETCHED';

/** IRS_PLAN_FETCHED action type */
export const REMOVE_IRS_PLANS = 'reveal/reducer/IRS/IRSPlan/REMOVE_IRS_PLANS';

/** IRS_PLAN_FETCHED action type */
export const ADD_IRS_PLAN = 'reveal/reducer/IRS/IRSPlan/ADD_IRS_PLAN';

/** interface for fetch IRSPlans action */
interface FetchIRSPlansAction extends AnyAction {
  IRSPlansById: { [key: string]: IRSPlan };
  type: typeof IRS_PLANS_FETCHED;
}

/** interface for removing IRSPlans action */
interface RemoveIRSPlansAction extends AnyAction {
  IRSPlansById: { [key: string]: IRSPlan };
  type: typeof REMOVE_IRS_PLANS;
}

/** interface for adding a single IRSPlans action */
interface AddIRSPlanAction extends AnyAction {
  IRSPlanObj: IRSPlan;
  type: typeof ADD_IRS_PLAN;
}

/** Create type for IRSPlan reducer actions */
export type IRSPlanActionTypes =
  | FetchIRSPlansAction
  | RemoveIRSPlansAction
  | AddIRSPlanAction
  | AnyAction;

// action creators

/**
 * Fetch Plan Definitions action creator
 * @param {IRSPlan[]} IRSPlanList - list of plan definition objects
 */
export const fetchIRSPlans = (IRSPlanList: IRSPlan[] = []): FetchIRSPlansAction => ({
  IRSPlansById: keyBy(IRSPlanList, 'plan_id'),
  type: IRS_PLANS_FETCHED,
});

/** Reset plan definitions state action creator */
export const removeIRSPlans = () => ({
  IRSPlansById: {},
  type: REMOVE_IRS_PLANS,
});

/**
 * Add one Plan Definition action creator
 * @param {IRSPlan} IRSPlanObj - the plan definition object
 */
export const addIRSPlan = (IRSPlanObj: IRSPlan): AddIRSPlanAction => ({
  IRSPlanObj,
  type: ADD_IRS_PLAN,
});

// the reducer

/** interface for IRSPlan state */
interface IRSPlanState {
  IRSPlansById: { [key: string]: IRSPlan };
}

/** immutable IRSPlan state */
export type ImmutableIRSPlanState = IRSPlanState & SeamlessImmutable.ImmutableObject<IRSPlanState>;

/** initial IRSPlan state */
const initialState: ImmutableIRSPlanState = SeamlessImmutable({
  IRSPlansById: {},
});

/** the IRSPlan reducer function */
export default function reducer(
  state = initialState,
  action: IRSPlanActionTypes
): ImmutableIRSPlanState {
  switch (action.type) {
    case IRS_PLANS_FETCHED:
      if (action.IRSPlansById) {
        return SeamlessImmutable({
          ...state,
          IRSPlansById: { ...state.IRSPlansById, ...action.IRSPlansById },
        });
      }
      return state;
    case ADD_IRS_PLAN:
      if (action.IRSPlanObj as IRSPlan) {
        return SeamlessImmutable({
          ...state,
          IRSPlansById: {
            ...state.IRSPlansById,
            [action.IRSPlanObj.plan_id as string]: action.IRSPlanObj,
          },
        });
      }
      return state;
    case REMOVE_IRS_PLANS:
      return SeamlessImmutable({
        ...state,
        IRSPlansById: action.IRSPlansById,
      });
    default:
      return state;
  }
}

// selectors

/** get IRSPlans by id
 * @param {Partial<Store>} state - the redux store
 * @param {InterventionType} interventionType - the IRSPlan intervention Type
 * @returns {{ [key: string]: IRSPlan }} IRSPlans by id
 */
export function getIRSPlansById(
  state: Partial<Store>,
  interventionType: InterventionType = InterventionType.IRS
): { [key: string]: IRSPlan } {
  if (interventionType) {
    return keyBy(getIRSPlansArray(state, interventionType), 'plan_id');
  }
  return (state as any)[reducerName].IRSPlansById;
}

/** get one IRSPlan using its id
 * @param {Partial<Store>} state - the redux store
 * @param {string} planId - the IRSPlan id
 * @returns {IRSPlan|null} a IRSPlan object or null
 */
export function getIRSPlanById(state: Partial<Store>, planId: string): IRSPlan | null {
  return get((state as any)[reducerName].IRSPlansById, planId) || null;
}

/** get an array of IRSPlan objects
 * @param {Partial<Store>} state - the redux store
 * @param {InterventionType} interventionType - the IRSPlan intervention Type
 * @returns {IRSPlan[]} an array of IRSPlan objects
 */
export function getIRSPlansArray(
  state: Partial<Store>,
  interventionType: InterventionType = InterventionType.IRS
): IRSPlan[] {
  const result = values((state as any)[reducerName].IRSPlansById);
  return result.filter((e: IRSPlan) => e.plan_intervention_type === interventionType);
}
