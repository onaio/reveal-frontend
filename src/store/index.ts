import { store } from '@opensrp/store';

export default store;
