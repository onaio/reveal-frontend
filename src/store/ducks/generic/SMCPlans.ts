import {
  fetchActionCreatorFactory,
  getItemByIdFactory,
  getItemsByIdFactory,
  reducerFactory,
  removeActionCreatorFactory,
} from '@opensrp/reducer-factory';
import intersect from 'fast_array_intersect';
import { values } from 'lodash';
import { Store } from 'redux';
import { createSelector } from 'reselect';
import { GenericPlan } from './plans';

/** the reducer name */
export const reducerName = 'SMCPlans';

/** generic plans Reducer */
const reducer = reducerFactory<GenericPlan>(reducerName);
export default reducer;

// actions
/** actionCreator returns action to to add Item records to store */
export const fetchSMCPlans = fetchActionCreatorFactory<GenericPlan>(reducerName, 'plan_id');
/** actionCreator returns action to to remove Item records to store */
export const removeSMCPlans = removeActionCreatorFactory(reducerName);

// selectors
/** get one Plan using its id */
export const getSMCPlanById = getItemByIdFactory<GenericPlan>(reducerName);

/** get all plans by ids */
export const getSMCPlansArray = getItemsByIdFactory<GenericPlan>(reducerName);

/** RESELECT USAGE STARTS HERE */

/** This interface represents the structure of SMC plan filter options/params */
export interface SMCPlanFilters {
  plan_title?: string /** SMC plan title */;
  statusList?: string[] /** array of plan statuses */;
}

/** SMCPlansArrayBaseSelector select an array of all plans
 * @param state - the redux store
 */
export const SMCPlansArrayBaseSelector = (state: Partial<Store>): GenericPlan[] =>
  values(getSMCPlansArray(state) || {});

/** getSMCPlansArrayByTitle
 * Gets title from PlanFilters
 * @param state - the redux store
 * @param props - the plan filters object
 */
export const getTitle = (_: Partial<Store>, props: SMCPlanFilters) => props.plan_title;

/** getStatusList
 * Gets statusList from PlanFilters
 * @param state - the redux store
 * @param props - the plan filters object
 */
export const getStatusList = (_: Partial<Store>, props: SMCPlanFilters) => props.statusList;

/** getPlansArrayByTitle
 * Gets an array of Plan objects filtered by plan title
 * @param {Partial<Store>} state - the redux store
 * @param {PlanDefinitionFilters} props - the plan defintion filters object
 */
export const getSMCPlansArrayByTitle = () =>
  createSelector([SMCPlansArrayBaseSelector, getTitle], (plans, title) =>
    title
      ? plans.filter(plan => plan.plan_title.toLowerCase().includes(title.toLowerCase()))
      : plans
  );

/** getSMCPlansArrayByStatus
 * Gets an array of Plan objects filtered by plan status
 * @param {Partial<Store>} state - the redux store
 * @param {PlanDefinitionFilters} props - the plan defintion filters object
 */
export const getSMCPlansArrayByStatus = () =>
  createSelector([SMCPlansArrayBaseSelector, getStatusList], (plans, statusList) =>
    statusList
      ? plans.filter(plan => (statusList.length ? statusList.includes(plan.plan_status) : true))
      : plans
  );

/** makeSMCPlansArraySelector
 * Returns a selector that gets an array of SMCPlan objects filtered by one or all
 * of the following:
 *    - plan_title
 *
 * These filter params are all optional and are supplied via the prop parameter.
 *
 * This selector is meant to be a memoized replacement for getSMCPlansArray.
 *
 * To use this selector, do something like:
 *    const SMCPlansArraySelector = makeSMCPlansArraySelector();
 *
 * @param {Partial<Store>} state - the redux store
 * @param {PlanFilters} props - the plan filters object
 * @param {string} sortField - sort by field
 */
export const makeSMCPlansArraySelector = () => {
  return createSelector([getSMCPlansArrayByTitle(), getSMCPlansArrayByStatus()], (plan1, plan2) =>
    intersect([plan1, plan2], JSON.stringify)
  );
};
