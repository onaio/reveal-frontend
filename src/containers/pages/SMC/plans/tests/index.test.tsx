import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { createBrowserHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import { REPORT_SMC_PLAN_URL } from '../../../../../constants';
import store from '../../../../../store';
import { fetchSMCPlans, SMCPLANType } from '../../../../../store/ducks/generic/SMCPlans';
import * as fixtures from '../../../../../store/ducks/generic/tests/fixtures';
import ConnectedSMCPlansList, { SMCPlansList } from '../index';

jest.mock('../../../../../configs/env');
/* tslint:disable-next-line no-var-requires */
const fetch = require('jest-fetch-mock');

const history = createBrowserHistory();

const props = {
  history,
  location: {
    hash: '',
    pathname: REPORT_SMC_PLAN_URL,
    search: '',
    state: undefined,
  },
  match: {
    isExact: true,
    params: {},
    path: `${REPORT_SMC_PLAN_URL}/`,
    url: `${REPORT_SMC_PLAN_URL}/`,
  },
};

describe('components/SMC/SMCPlansList', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    shallow(
      <Router history={history}>
        <SMCPlansList {...props} />
      </Router>
    );
  });

  it('renders plan definition list correctly', () => {
    store.dispatch(fetchSMCPlans(fixtures.SMCPlans as SMCPLANType[]));
    fetch.mockResponseOnce(JSON.stringify({}));
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedSMCPlansList {...props} />
        </Router>
      </Provider>
    );
    expect(toJson(wrapper.find('BreadcrumbItem li'))).toMatchSnapshot('breadcrumbs');
    expect(wrapper.find('h3.page-title').text()).toEqual('SMC Plans');
    const tableHeader = wrapper.find('.thead .tr .th');
    Array(tableHeader.length)
      .fill('')
      .forEach((_, i) => {
        expect(tableHeader.at(i).text()).toMatchSnapshot(`table header-${i + 1}`);
      });
    expect(wrapper.find('.tbody .tr').length).toEqual(3);
    // have searchbar
    expect(wrapper.find('.search-input-wrapper').length).toEqual(1);
    // have top and bottom pagination
    expect(wrapper.find('.pagination').length).toEqual(2);
    // have height and columns filters
    expect(
      wrapper
        .find('.filter-bar-btns span')
        .at(0)
        .text()
    ).toEqual('Row Height');
    expect(
      wrapper
        .find('.filter-bar-btns span')
        .at(1)
        .text()
    ).toEqual('Customize Columns');

    expect(wrapper.find('GenericPlansList').length).toBe(1);
    expect(wrapper.find('GenericPlansList').props()).toMatchSnapshot('GenericPlansList props');

    expect(wrapper.find('.page-title').text()).toEqual('SMC Plans');

    wrapper.unmount();
  });

  it('handles search correctly', async () => {
    store.dispatch(fetchSMCPlans(fixtures.SMCPlans as SMCPLANType[]));
    props.location.search = '?title=Cycle 4';
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedSMCPlansList {...props} />
        </Router>
      </Provider>
    );
    expect(wrapper.find('.tbody .tr .td a').text()).toEqual(
      'Nigeria Cycle 4 - SMC Implementation Plan'
    );

    expect((wrapper.find('GenericPlansList').props() as any).plans).toEqual([fixtures.SMCPlans[1]]);
  });
});
