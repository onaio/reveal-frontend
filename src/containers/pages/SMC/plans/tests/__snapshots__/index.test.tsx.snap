// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/SMC/SMCPlansList renders plan definition list correctly: GenericPlansList props 1`] = `
Object {
  "fetchPlans": [Function],
  "history": Object {
    "action": "POP",
    "block": [Function],
    "createHref": [Function],
    "go": [Function],
    "goBack": [Function],
    "goForward": [Function],
    "length": 1,
    "listen": [Function],
    "location": Object {
      "hash": "",
      "pathname": "/",
      "search": "",
      "state": undefined,
    },
    "push": [Function],
    "replace": [Function],
  },
  "location": Object {
    "hash": "",
    "pathname": "/intervention/smc/report",
    "search": "",
    "state": undefined,
  },
  "match": Object {
    "isExact": true,
    "params": Object {},
    "path": "/intervention/smc/report/",
    "url": "/intervention/smc/report/",
  },
  "pageTitle": "SMC Plans",
  "pageUrl": "/intervention/smc/report",
  "plans": Array [
    Object {
      "jurisdiction_root_parent_ids": Array [
        "12eaba32-55bb-4a67-b554-a28d9ab8e02a",
      ],
      "plan_date": "2020-12-04",
      "plan_effective_period_end": "2021-07-31",
      "plan_effective_period_start": "2020-12-04",
      "plan_id": "82d7d7fe-7897-5c69-a473-286cd6a9e0fb",
      "plan_intervention_type": "Dynamic-MDA",
      "plan_name": "Dynamic-MDA-2020-12-04",
      "plan_status": "active",
      "plan_title": "Dynamic-MDA 2020-12-04_2030",
      "plan_version": "1",
    },
    Object {
      "jurisdiction_root_parent_ids": Array [
        "12eaba32-55bb-4a67-b554-a28d9ab8e02a",
      ],
      "plan_date": "2020-10-09",
      "plan_effective_period_end": "2021-09-09",
      "plan_effective_period_start": "2020-10-09",
      "plan_id": "12a879d6-8ad8-4198-86c7-0e94f0dcc848",
      "plan_intervention_type": "MDA",
      "plan_name": "Nigeria-Cycle4-SMC-Implementation-Plan",
      "plan_status": "active",
      "plan_title": "Nigeria Cycle 4 - SMC Implementation Plan",
      "plan_version": "2",
    },
    Object {
      "jurisdiction_root_parent_ids": Array [
        "12eaba32-55bb-4a67-b554-a28d9ab8e02a",
      ],
      "plan_date": "2020-09-08",
      "plan_effective_period_end": "2021-09-09",
      "plan_effective_period_start": "2020-09-08",
      "plan_id": "262526a3-fe86-4cb5-a743-f26db7ff3ce0",
      "plan_intervention_type": "MDA",
      "plan_name": "Nigeria-Cycle3-SMC-Implementation-Plan",
      "plan_status": "draft",
      "plan_title": "Nigeria Cycle 3 - SMC Implementation Plan",
      "plan_version": "1",
    },
  ],
  "service": [Function],
  "supersetReportingSlice": "1",
}
`;

exports[`components/SMC/SMCPlansList renders plan definition list correctly: breadcrumbs 1`] = `
Array [
  <li
    className="breadcrumb-item"
  >
    <Link
      key="0"
      to="/"
    >
      <LinkAnchor
        href="/"
        navigate={[Function]}
      >
        <a
          href="/"
          onClick={[Function]}
        >
          Home
        </a>
      </LinkAnchor>
    </Link>
  </li>,
  <li
    aria-current="page"
    className="active breadcrumb-item"
  >
    SMC Plans
  </li>,
]
`;

exports[`components/SMC/SMCPlansList renders plan definition list correctly: table header-1 1`] = `"Title"`;

exports[`components/SMC/SMCPlansList renders plan definition list correctly: table header-2 1`] = `"Date Created"`;

exports[`components/SMC/SMCPlansList renders plan definition list correctly: table header-3 1`] = `"Start Date"`;

exports[`components/SMC/SMCPlansList renders plan definition list correctly: table header-4 1`] = `"End Date"`;

exports[`components/SMC/SMCPlansList renders plan definition list correctly: table header-5 1`] = `"Status"`;
